Show if-case and for-case in the guide (#361)
swiftlang/swift-book: TSPL.docc/LanguageGuide/ControlFlow.md

which means that the code in the body of the `case`
can always access a value for `distance`.

> ## Patterns

Author: Reviewers: Open to alternate names for this section.  This is a bit of a placeholder heading.

Author: @colleenish Any suggestions or opinions on "Patterns" as a subheading?

In the previous examples, each switch case includes a pattern
that indicates what values match that case.
You can also use a pattern as the condition for an `if` statement.
Here's what that looks like:

```swift
let somePoint = (12, 100)
if case (let x, 100) = somePoint {
    print("Found a point on the y=100 line, at \(x)")
}
// Prints "Found a point on the y=100 line, at 12"
```

In this code,
the condition for the `if` statement starts with `case`,
indicating that the condition is a pattern instead of a Boolean value.
If the pattern matches,
the condition for the `if` is considered to be true,

> and the code in the body of the `if` statement runs.

Reviewer: Should "and the code in the body" be "then the code in the body", to make it clear that it's a consequence rather than a third condition?

The patterns you can write after `if case`
are the same as the patterns you can write in a switch case.

You can also use a pattern in a `for`-`in` loop,
to give names to parts of the value using a value binding:

```swift
let points = [(10, 0), (30, 100), (-20, 0)]

for case let (x, y) in points {
    if y == 0 {
        print("Found a point on the x-axis at \(x)")
    }
}
// Prints "Found a point on the x-axis at 10"
// Prints "Found a point on the x-axis at -20"

> ```

Reviewer: This example doesn't need the case at, i.e., you can write:

```swift
  for (x, y) in points {
    if y == 0 {
      print("Found a point on the x-axis at \(x)")
    }
  }
```

If you want to match y == 0, you can use this as an example:

```swift
  for case (let x, 0) in points {
    print("Found a point on the x-axis at \(x)")
  }
```

The `for`-`in` loop above iterates over an array of tuples,
binding the first and second elements of the tuples
to the `x` and `y` constants.
The statements inside the loop can use those constants,
like the `if` statement that checks whether the point lies on the x-axis.

A `for`-`case`-`in` loop can also include a `where` clause,
to check for an additional condition.
The statements inside the loop run
only when `where` clause matches the current element.
For example,
the `for`-`case`-`in` loop below is the same as the `for`-`in` loop above.

```swift
for case let (x, y) in points where y == 0  {
    print("Found a point on the x-axis at \(x)")
}
// Prints "Found a point on the x-axis at 10"
// Prints "Found a point on the x-axis at -20"

> ```

Reviewer: This example doesn't need the `case`, either. What if you make this part about "any `for`-`in` loop" having a `where` clause, and capture points along a line that goes through the origin?

```swift
for case let (x, y) in points where x == y || x == -y  {
    print("Found a point (\(x), \(y)) along line through the origin")
}
```

which means that the code in the body of the `case`
can always access a value for `distance`.

## Patterns

In the previous examples, each switch case includes a pattern
that indicates what values match that case.
You can also use a pattern as the condition for an `if` statement.
Here's what that looks like:

```swift
let somePoint = (12, 100)
if case (let x, 100) = somePoint {
    print("Found a point on the y=100 line, at \(x)")
}

> // Prints "Found a point on the y=100 line, at 12"

Reviewer 2: ```suggestion
// Prints "Found a point on the y=100 line, at 12".
```

Author: TSPL consistently omits the period in this position.  This was an intentional change in 2015, at the request of Editorial, because the period next to a quotation mark is too busy.  If we want to switch back, we should track that separately — and it will have to be a manual process now, since the comments like this are no longer generated by a code testing system like they were at that time.

(See commits 7bbdce02ec8b75f1, a3f3d8ee2924b8a5, and c581094331f4b47a in the internal repo, from before this project was migrated to open source.)

```

In this code,
the condition for the `if` statement starts with `case`,
indicating that the condition is a pattern instead of a Boolean value.
If the pattern matches,
then the condition for the `if` is considered to be true,
and so the code in the body of the `if` statement runs.
The patterns you can write after `if case`
are the same as the patterns you can write in a switch case.

In a `for`-`in` loop,
you can give names to parts of the value using a value binding pattern,
even without writing `case` in your code:

```swift
let points = [(10, 0), (30, -30), (-20, 0)]

for (x, y) in points {
    if y == 0 {
        print("Found a point on the x-axis at \(x)")
    }
}

> // Prints "Found a point on the x-axis at 10"

Reviewer 2: ```suggestion
// Prints "Found a point on the x-axis at 10".
```

> // Prints "Found a point on the x-axis at -20"

Reviewer 2: ```suggestion
// Prints "Found a point on the x-axis at -20".
```

```

The `for`-`in` loop above iterates over an array of tuples,
binding the first and second elements of the tuples
to the `x` and `y` constants.
The statements inside the loop can use those constants,

> like the `if` statement that checks whether the point lies on the x-axis.

Reviewer 2: ```suggestion
such as the `if` statement that checks whether the point lies on the x-axis.
```

A more concise way to write this code

> is to combine the value bindings and condition,

Reviewer 2: ```suggestion
is to combine the value bindings and condition
```

using a `for`-`case`-`in` loop.
The code below has the same behavior as the `for`-`in` loop above:

```swift
for case (let x, 0) in points {
    print("Found a point on the x-axis at \(x)")
}

> // Prints "Found a point on the x-axis at 10"

Reviewer 2: ```suggestion
// Prints "Found a point on the x-axis at 10".
```

> // Prints "Found a point on the x-axis at -20"

Reviewer 2: ```suggestion
// Prints "Found a point on the x-axis at -20".
```

```

In this code,
the condition is integrated into the `for`-`case`-`in` loop
as part of the pattern.
The statements in the `for`-`case`-`in` loop run only for points on the x-axis.
This code produces the same result as the `for`-`in` loop above,
but is a more compact way to iterate
over only certain elements in a collection.

A `for`-`case`-`in` loop can also include a `where` clause,
to check for an additional condition.
The statements inside the loop run
only when `where` clause matches the current element.
For example:

```swift
for case let (x, y) in points where x == y || x == -y  {
    print("Found (\(x), \(y)) along a line through the origin")
}

> // Prints "Found (30, -30) along a line through the origin"

Reviewer 2: ```suggestion
// Prints "Found (30, -30) along a line through the origin".
```
